### PR TITLE
WIP: Add first basic working prototype of Token class

### PR DESCRIPTION
**Pull request by norpol**
Work in progress...
Going to update this description when I feel the code is more ready.
#20 

- [ ] add `refresh_token`
- [ ] add `grant_type`
- [ ] fix `token.read()` should remove inexistent tokens
- [ ] cover APIs requirement of what is exclusive
- [ ] add tests
- [ ] add documentation
- [ ] update docstrings
- [ ] maybe make PyJWT `verify` configurable (though, I don't know how the tokens are signed tbh)

```python3
# initialize token instance
scope = "member-of-groups: readers"
token = Token(artifactory, scope=scope, expires_in=3600)

# create token
token.create()

# get jwt_token alone
token.jwt_token

# get Artifactory API json response as dict
token.token

# get dictionary of all tokens for a certain scope
token.read()
token.tokens

# delete all those tokens
for key in token.tokens:
    t = Token(artifactory, token_id=key)
    t.delete()

# should not show any other tokens
token = Token(artifactory, scope=scope, expires_in=3600)
token.read()
token.tokens
```

### CODE REVIEW

**Comment by Arlington1985** [comment]
Any deadline for this release? I really need it maybe I can help on that

**Comment by norpol** [comment]
I stopped working at the company that was using this code, the core dev is quite responsive. My coworker tried to take the pull-request over, but apparently also never finished:
https://github.com/devopshq/artifactory/pull/41

I don't have access to Artifactory Pro anymore, the code is basically working - but I think for merging into master it needs some cleanups improvements (as stated in my TODO above).

Do you want to take it over @Arlington1985? cc @thaligar.

File: dohq_artifactory/admin.py [hunk 1]
```diff
@@ -418,3 +418,147 @@ def add_group(self, name, permissions):
...
+        grant_type=None,
+    ):
+        super(Token, self).__init__(artifactory)
+
+        # Either or is optional
+        if not (username or scope):
+            raise TypeError("Require either username or scope as argument")
+
+        if username is None:
+            username = self._artifactory.session.auth[0]
```
Comment (Arlington1985) [inline dohq_artifactory/admin.py]: This part failed for me. Instead, I suggest that one:
`username = self._artifactory.session.auth.username`

**Comment by Arlington1985** [comment]
Also we have two commits for the same change, which one is relevant and going to merge?

- https://github.com/devopshq/artifactory/pull/21/commits/46abb5629c61cd5d386337c8e662e40e956ba72c
- https://github.com/devopshq/artifactory/pull/41/commits/96daa628bd6df7d463df7b25fce2c9d399581085

**Comment by norpol** [comment]
I guess my coworker did just a rebase then.

**Comment by Arlington1985** [comment]
Should I create third merge request ?

**Comment by norpol** [comment]
I think that would make sense, I'll close mine in favor of yours then.

Thank you for your work! Artifactory is by far not ideal, but its the best tool that I'm aware of and a Python library is really important...

I wish Artifactory was an opensource community project though.

**Comment by Arlington1985** [comment]
@norpol , what you mean here **fix token.read() should remove inexistent tokens** ? Why read should delete anything?

**Comment by norpol** [comment]
I'm not sure anymore, quick thing I have in mind is something like you create a token, then delete a token and then run `token.read()` its still shown as a token maybe?

**Comment by norpol** [comment]
If this doesn't make sense I can try to review the code once more.

File: dohq_artifactory/admin.py (hunk 1 above)
Comment (allburov) [reply dohq_artifactory/admin.py]: I suggest fail if `username is None`. The library support many type of Auth, it's hard to support all of them. https://github.com/devopshq/artifactory#authentication

**Comment by allburov** [comment]
>Should I create third merge request ?

@Arlington1985 I think you should :)  You could copy TODOs from this issue

**Comment by allburov** [comment]
Closed in favor of : https://github.com/devopshq/artifactory/pull/103

**Pull request closed without merging**